feat: chatty ui

kyteinsky opened the pull request.

julien-nc reviewed (changes requested):
Incredibly nice!
It looks awesome and works super good. Congrats! :muscle: 
I had forgotten you were using the `textProcessingManager->runTask` so it was such a nice feeling to immediately get the answer after sending my first message. :grin: 
The instruct being in english does not prevent gpt-3.5-turbo to answer in the same language as the last user message :tada: I guess/hope it's the same for most LLMs.

Some UI findings:
* The chattyUI expands too much vertically, making the entire `modal-container__content` overflow and scroll. 
![image](https://github.com/nextcloud/assistant/assets/11291457/4299a0e8-1c4a-41e5-b8f7-d04c2195e81b)

* It would be nice to have a visual feedback when the LLM answer is being computed. How about a simple loading spinner (NcLoadingIcon for example) where the next message will appear?
* Something goes wrong when regenerating. The last user message (before the assistant message to regenerate) is deleted (from the UI only). It comes back after switching sessions or reloading. This does not happen if one asks to regenerate a message right after loading the session. I can give you more details if needed.
* The instruct prefix is lost when chat_last_n_messages is exceeded. See inline comment in the controller.

marcelklehr commented on appinfo/routes.php:
```diff
@@ -24,6 +24,17 @@
+		['name' => 'chattyLLM#getMessages', 'url' => '/chat/messages', 'verb' => 'GET'],
+		['name' => 'chattyLLM#generate', 'url' => '/chat/generate', 'verb' => 'GET'],
+		['name' => 'chattyLLM#regenerate', 'url' => '/chat/regenerate', 'verb' => 'GET'],
+		['name' => 'chattyLLM#generateTitle', 'url' => '/chat/generate_title', 'verb' => 'GET'],
```
At some point it would be nice to migrate to the new Route annotation  by kate

kyteinsky replied:
We would be losing support for the previous server versions if we do that now. Also, personally I would like to keep this alongside that, if possible, to have a quick look at all the endpoints declared in the app.

marcelklehr commented on lib/Controller/ChattyLLMController.php:
```diff
@@ -0,0 +1,362 @@
+		}
+
+		try {
+			$this->sessionMapper->deleteSession($sessionId);
```
QueryMapper exposes a delete method already, so you could just 
```suggestion
			$session = $this->sessionMapper->find($sessionId);
			$this->sessionMapper->delete($session);
```

kyteinsky replied:
Cool! I was missing the find method in my thought process.

kyteinsky replied:
looks like a query has to be built for find anyway so this way should be fine ig?

marcelklehr commented on lib/Settings/Admin.php:
```diff
@@ -37,6 +39,9 @@ public function getForm(): TemplateResponse {
 		$freePromptPickerEnabled = $this->config->getAppValue(Application::APP_ID, 'free_prompt_picker_enabled', '1') === '1';
 		$speechToTextAvailable = $this->speechToTextManager->hasProviders();
 		$speechToTextEnabled = $this->config->getAppValue(Application::APP_ID, 'speech_to_text_picker_enabled', '1') === '1';
+		$chattyLLMUserInstructions = $this->config->getAppValue(Application::APP_ID, 'chat_user_instructions', $this->l10n->t(Application::CHAT_USER_INSTRUCTIONS));
+		$chattyLLMUserInstructionsTitle = $this->config->getAppValue(Application::APP_ID, 'chat_user_instructions_title', $this->l10n->t(Application::CHAT_USER_INSTRUCTIONS_TITLE));
```
Shall we just leave the instructions in English and add an instruction to take over the language of the user?

kyteinsky replied:
weird since the string should be put in place by php and shouldn't have to do anything with the function call. By "source parser" you mean it is all raw code parsing to get the list of all translatable strings?

marcelklehr replied:
Yeah, it's using regular expressions ;)

kyteinsky replied:
:grimacing:

kyteinsky replied:
> Shall we just leave the instructions in English and add an instruction to take over the language of the user?

Yeah that could be done but different models behave differently with this. I tried asking something in German when the user instruction was in English, it responded in German but explained it in English as well.

But I still think your approach is the way to go. Admins can change the instructions to their liking afterwards.

marcelklehr replied:
Please remove the ->t calls.

kyteinsky commented on src/components/AdminSettings.vue:
```diff
@@ -82,36 +82,95 @@
+					:link-auto-complete="false"
+					:placeholder="t('assistant', 'Chat User Instructions')"
+					:aria-label="t('assistant', 'Chat User Instructions')"
+					dir="auto"
```
yeah, tested right now, it is required.

You're right, it should be set by default.

kyteinsky commented on src/components/AdminSettings.vue:
```diff
@@ -82,36 +82,95 @@
+					:placeholder="t('assistant', 'Chat User Instructions')"
+					:aria-label="t('assistant', 'Chat User Instructions')"
+					dir="auto"
+					@update:value="delayedValueUpdate(state.chat_user_instructions, 'chat_user_instructions')"
```
It is already debounced though.

```js
		delayedValueUpdate(newValue, key) {
			this.optionsToSave[key] = newValue

			if (this.textFieldSaveTimer) {
				clearTimeout(this.textFieldSaveTimer)
			}
			this.textFieldSaveTimer = setTimeout(() => {
				this.saveOptions(this.optionsToSave)
			}, 2000)
		},
```

marcelklehr replied:
I think julien means you can save some code by using the 'debounce' package

marcelklehr replied:
makes it easier to read, too

kyteinsky replied:
I guess it would increase the amount of code if we use a package and doesn't really have much impact on functionality here imo. We can extract it into a util like this, still keeping it local: https://github.com/nextcloud/integration_mattermost/blob/main/src/utils.js#L8-L18

julien-nc replied:
Yeah fine if you prefer using your own debounce implementation. Just put it in a util file.

kyteinsky commented on src/components/ChattyLLM/ChattyLLMInputForm.vue:
```diff
@@ -0,0 +1,645 @@
+					{{ t('assistant', 'Loading sessions...') }}
+				</div>
+				<div v-else-if="sessions != null && sessions.length === 0" class="unloaded-sessions">
+					{{ t('assistant', 'So empty') }}
```
NcEmptyContent just changes the font size when in "name" and the color when in "description". What is the issue with the current thing?

julien-nc replied:
At least change the text. Instead of "So empty" you could have "No session yet".

kyteinsky replied:
done :+1:

kyteinsky commented on lib/Controller/ChattyLLMController.php:
```diff
@@ -0,0 +1,362 @@
+	 * @return JSONResponse
+	 */
+	#[NoAdminRequired]
+	public function regenerate(int $sessionId, int $messageId): JSONResponse {
```
> How about "regenerateAnswerFromMessage"?

I am looking at the generation from a session perspective. An answer is generated for a chat session rather than the last message. We stitch the instructions and the messages, append "assistant: " to it and let the llm generate a response based on that. There is no if-else needed here since the llm is smart enough to figure it out, if not the user is always there to correct it.

The messageId here is confusing since that only refers to the point from where all the newer history is to be wiped. It is constrained in the UI to only allow assistant messages to be the origins of destruction but the delete feature will allow the users to delete user messages leaving two consecutive assistant messages. I honestly think this is fine and shouldn't be disallowed with more conditionals which would then be confusing for the end user and developers alike.

kyteinsky commented on lib/Controller/ChattyLLMController.php:
```diff
@@ -0,0 +1,362 @@
+
+			$message = new Message();
+			$message->setSessionId($sessionId);
+			$message->setRole($this->l10n->t('assistant'));
```
It will be needed for the final prompt since there might be a model that does not understand English well or on the opposite end, might be biased towards English and might not reply in the asked language.

But we should set it in English in the db.

kyteinsky replied:
following this (https://github.com/nextcloud/assistant/pull/86#discussion_r1598250791), it would make sense to leave the roles untranslated everywhere.

julien-nc commented:
Additional requirement: Mention in the README the length of the assistant memory (number of messages used to generate a new one). Mention it can be configured with an app config value.

kyteinsky commented:
> It would be nice to have a visual feedback when the LLM answer is being computed. How about a simple loading spinner (NcLoadingIcon for example) where the next message will appear?

The input box shows the "Thinking" text and there is a breathing animation as well. Maybe it needs to be more prominent. Do you suggest a loading icon in addition to this?

> Something goes wrong when regenerating. The last user message (before the assistant message to regenerate) is deleted (from the UI only). It comes back after switching sessions or reloading. This does not happen if one asks to regenerate a message right after loading the session. I can give you more details if needed.

Fixed.

> The instruct prefix is lost when chat_last_n_messages is exceeded. See inline comment in the controller.

Fixed.

julien-nc reviewed (commented):
A few more remarks:
* Maybe we can find something better than "session" like "conversation" (which you already use in the NoSession empty content)
* When clicking on "new session", it's not clear that one has to send a first message and THEN a new navigation entry will appear. How about showing a fake "new session" nav item when clicking on "new session"? But then it's not coherent with the initial state (no session selected, empty content displayed)... How about making it like in Talk? 
    * Initial empty content says "Join a conversation or start a new one"
    * Click on new session creates a nav entry and the empty content becomes "Hello there! What can I help you with today?"
    * So the session would be created before the first message

marcelklehr commented:
> There's no admin docs though, should I create one or just add it to the README in the root dir?

Could be done somewhere on this page? https://docs.nextcloud.com/server/latest/admin_manual/ai/app_assistant.html

marcelklehr commented:
> I guess/hope it's the same for most LLMs.


It isn't, I think.

marcelklehr commented on docs/admin/README.md:
````diff
@@ -0,0 +1,76 @@
+```
+occ config:app:set assistant chat_last_n_messages --value=10
+```
+
````
Please put all of this in the server admin docs under AI/Nextcloud Assistant to avoid having two sources of truth

marcelklehr commented on lib/AppInfo/Application.php:
```diff
@@ -63,6 +63,9 @@ class Application extends App implements IBootstrap {
 	public const TASK_CATEGORY_TEXT_TO_IMAGE = 1;
 	public const TASK_CATEGORY_SPEECH_TO_TEXT = 2;
 
+	public const CHAT_USER_INSTRUCTIONS = 'This is a conversation between {user} and you, Nextcloud Assistant. You are a kind, polite and helpful AI that helps {user} to the best of its abilities. If you do not understand something, you will ask for clarification. Remember to verify the premise presented by {user}. In a rare case, it might be wrong.';
```
```suggestion
	public const CHAT_USER_INSTRUCTIONS = 'This is a conversation in a specific language between {user} and you, Nextcloud Assistant in some langauge. You are a kind, polite and helpful AI that helps {user} to the best of its abilities. If you do not understand something, you will ask for clarification. Detect the language that {user} is using. Make sure to use the same language in your response. Do not mention the language explicitly. Remember to verify the premise presented by {user}. In a rare case, it might be wrong.';
```

marcelklehr replied:
some promp engineering to make it multilingual

marcelklehr commented on lib/AppInfo/Application.php:
```diff
@@ -63,6 +63,9 @@ class Application extends App implements IBootstrap {
 	public const TASK_CATEGORY_SPEECH_TO_TEXT = 2;
 
+	public const CHAT_USER_INSTRUCTIONS = 'This is a conversation between {user} and you, Nextcloud Assistant. You are a kind, polite and helpful AI that helps {user} to the best of its abilities. If you do not understand something, you will ask for clarification. Remember to verify the premise presented by {user}. In a rare case, it might be wrong.';
+	public const CHAT_USER_INSTRUCTIONS_TITLE = 'Above is a chat session between {user} and you, Nextcloud Assistant. Generate a suitable title summarizing the conversation and output only that.';
```
```suggestion
	public const CHAT_USER_INSTRUCTIONS_TITLE = 'Above is a chat session in a specific language between {user} and you, Nextcloud Assistant. Generate a suitable title summarizing the conversation in the same language and output only that.';
```

marcelklehr reviewed (approved):
Unless julien has more comments, I'd say it's good to go

kyteinsky commented:
There are a few more tweaks to go, proposed by the design team.

julien-nc commented on src/components/ChattyLLM/ChattyLLMInputForm.vue:
```diff
@@ -0,0 +1,674 @@
+					@update:name="(newTitle) => onEditSessionTitle(session.id, newTitle)">
+					<template #actions>
+						<NcActionButton @click="onGenerateSessionTitle(session.id)">
+							<template v-if="!loading.titleGeneration" #icon>
+								<AutoFixIcon :size="20" />
+							</template>
+							<template v-else #icon>
+								<NcLoadingIcon :size="20" />
+							</template>
```
Could you just put one icon template and have the v-if on the icon components?

julien-nc commented on src/components/ChattyLLM/ChattyLLMInputForm.vue:
```diff
@@ -0,0 +1,674 @@
+				<div v-else-if="sessions != null && sessions.length === 0" class="unloaded-sessions">
+					{{ t('assistant', 'No conversations yet') }}
+				</div>
+				<NcAppNavigationItem
```
It just happens when there is another NcAppNavigation in the page.

kyteinsky replied:
nope, pulled server master just now. Works on both chromium and firefox.

kyteinsky replied:
> It just happens when there is another NcAppNavigation in the page.

yes it does, let me see if z-index or something can do the trick

kyteinsky replied:
hmm, I don't even have the popover mounted in my DOM. Can't really point at what's going wrong.

susnux replied:
While using the `NcAppNavigation` outside of the `NcContent` and within a dialog might work it is not supported usage of the component.

But a real problem here: It **must not** be used more than once on a page, it will break and even create invalid HTML content of the page. This will also cause any accessibility tests on that page to fail.

Instead the expected solution would probably be either:
```html
<NcDialog ...>
  <template #navigation>
    <ul>
      <NcListItem ...>
      <!-- .... -->
    </ul>
  </template>
  <!-- Content -->
</NcDialog>
```

or

```html
<NcDialog ...>
  <your-content-wrapper>
    <ul class="your-navigation">
      <NcListItem ...>
      <!-- .... -->
    </ul>
    <!-- Other content -->
  </your-content-wrapper>
</NcDialog>
```

marcelklehr commented:
Super cool :tada:
